File: Assets/src/model/enemies/TelegraphedTask.cs
using System;
using System.Collections.Generic;

[System.Serializable]
public class TelegraphedTask : ActorTask {
  public override TaskStage WhenToCheckIsDone => TaskStage.After;
  private int turns;
  private readonly ActionType type;
  // temp
  public BaseAction then;
  private bool done;

  public TelegraphedTask(Actor actor, int turns, BaseAction then, ActionType type) : base(actor) {
    this.turns = turns;
    this.type = type;
    this.then = then;
  }
  public TelegraphedTask(Actor actor, int turns, BaseAction then) : this(actor, turns, then, then.Type) {}

  protected override BaseAction GetNextActionImpl() {
    // 3, 2, 1
    if (turns > 0) {
      turns--;
      return new WaitBaseAction(actor, type);
    }
    done = true;
    return then;
  }

  public override bool IsDone() => done;
}

File: web/src/model/tasks/TelegraphedTask.ts
import { ActorTask, TaskStage } from '../ActorTask';
import { WaitBaseAction, type BaseAction } from '../BaseAction';
import { ActionType } from '../../core/types';
import type { Actor } from '../Actor';

export class TelegraphedTask extends ActorTask {
  private turns: number;
  private readonly actionType: ActionType;
  then: BaseAction;
  private done = false;

  get whenToCheckIsDone(): TaskStage {
    return TaskStage.After;
  }

  constructor(actor: Actor, turns: number, then: BaseAction, actionType?: ActionType) {
    super(actor);
    this.turns = turns;
    this.then = then;
    this.actionType = actionType ?? then.type;
  }

  protected getNextActionImpl(): BaseAction {
    if (this.turns > 0) {
      this.turns--;
      return new WaitBaseAction(this.actor, this.actionType);
    }
    this.done = true;
    return this.then;
  }

  postStep(action: BaseAction, finalAction: BaseAction): void {
    if (action !== finalAction) {
      this.done = true;
    }
  }

  isDone(): boolean {
    return this.done;
  }
}
